dist/sizzle.js

```diff
@@ -1628,13 +1633,237 @@ function setFilters() {}
+		var rightValue = TOKEN_TYPES_VALUES[right.type];
+		return leftValue - rightValue;
+	};
+
+	/**
+	 * Checks if the specified tokens group is sortable.
+	 * We do not re-sort tokens in case of any positional or child pseudos in the group
+	 */
+	var isSortable = function(tokens) {
+
+		var iTokens = tokens.length;
+		while (iTokens--) {
+			var token = tokens[iTokens];
+			if (token.type === "PSEUDO" && POSITIONAL_PSEUDOS.indexOf(token.matches[0]) !== -1) {
+				return false;
+			}
+			if (token.type === "CHILD") {
+				return false;
+			}
+		}
+		return true;
+	};
+
+	/**
+	 * Sorts the tokens in order to mitigate the issues caused by the left-to-right matching.
+	 * The idea is change the tokens order so that Sizzle was matching fast selectors first (id, class),
+	 * and slow selectors after that (and here I mean our slow custom pseudo classes).
+	 * 
+	 * @param {Array} tokens An array of tokens to sort
+	 * @returns {Array} A new re-sorted array
+	 */
+	var sortTokens = function(tokens) {
+
+		if (!tokens || tokens.length === 1) {
+			return tokens;
+		}
+
+		var sortedTokens = [];
+		var groups = splitCompoundSelector(tokens);
+		for (var i = 0; i < groups.length; i++) {
+			var group = groups[i];
+			if (group instanceof Array) {
+				if (isSortable(group)) {
+					group.sort(compareFunction);
+				}
+				sortedTokens = sortedTokens.concat(group);
+			} else {
+				sortedTokens.push(group);
+			}
+		}
+
+		return sortedTokens;
+	};
+
+	/**
+	 * Sorts every tokens array inside of the specified "groups" array.
+	 * See "sortTokens" methods for more information on how tokens are sorted.
+	 * 
+	 * @param {Array} groups An array of tokens arrays.
+	 * @returns {Array} A new array that consists of the same tokens arrays after sorting
+	 */
+	var sort = function(groups) {
+		var sortedGroups = [];
+		var len = groups.length;
+		var i = 0;
+		for (; i < len; i++) {
+			sortedGroups.push(sortTokens(groups[i]));
+		}
+		return sortedGroups;
+	};
+
+	// Expose
+	return {
+		sort: sort
+	};
+})();
+
+/**
+ * [AdGuard Patch]:
+ * Removes trailing spaces from the tokens list
+ * 
+ * @param {*} tokens An array of Sizzle tokens to post-process
+ */
+function removeTrailingSpaces(tokens) {
+	var iTokens = tokens.length;
+	while (iTokens--) {
+		var token = tokens[iTokens];
+		if (token.type === " ") {
+			tokens.length = iTokens;
+		} else {
+			break;
+		}
+	}
+}
+
+/**
+ * [AdGuard Patch]:
+ * 
+ * This method processes parsed token groups, divides them into a number of selectors
+ * and makes sure that each selector's tokens are cached properly in Sizzle.
+ * 
+ * @typedef {Object} SelectorData
+ * @property {string} selectorText A CSS selector text
+ * @property {Array} groups An array of token groups corresponding to that selector
+ * 
+ * @param {*} groups Token groups (see {@link Sizzle.tokenize})
+ * @returns {Array.<SelectorData>} An array of selectors data we got from the groups
+ */
+function tolerantTokenize(groups) {
```

Comment:
The purpose of this function is unclear from the function name. This need to be changed to something like `tokensToSelectors`.

Reply:
++